lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,445 @@
+                if edgecase:
+                    exclude_self = node_edge_test(exclude_self, cube)
+                if not land[i, j] or not any(land[exclude_self]):
+                    continue
+
+                node_list = ConditionalListExtract('not_equal_to').process(
+                    land, node_list, 0)
+
+            dz_nearest = abs(altitude - orography[i, j])
+            dzs = altitude - orography[node_list]
+
+            dzs, dz_nearest, dz_subset = apply_bias(
+                relative_z, dzs, dz_nearest, altitude, orography, i, j)
+
+            ij_min = index_of_minimum_difference(dzs, subset_list=dz_subset)
+            i_min, j_min = list_entry_from_index(node_list, ij_min)
+            dz_min = abs(altitude - orography[i_min, j_min])
+
+            if dz_min < dz_nearest:
+                neighbour_list[i_site] = i_min, j_min, dzs[ij_min], edgecase
+
+        return neighbour_list
+
+# Wrapper routines to use the dz minimisation routine with various options.
+# These can be called as methods and set in the diagnostic configs.
+# It may be better to simply use the keyword options at a higher level,
+# but that will make the config more complex.
+
+    def min_dz_no_bias(self, cube, sites, **kwargs):
+        ''' Return local grid neighbour with minimum vertical displacement'''
+        return self.minimum_height_error_neighbour(cube, sites,
+                                                   relative_z=None,
+                                                   **kwargs)
+
+    def min_dz_biased_above(self, cube, sites, **kwargs):
+        '''
+        Return local grid neighbour with minimum vertical displacement,
+        biased to select grid points above the site altitude.
+
+        '''
+        return self.minimum_height_error_neighbour(cube, sites,
+                                                   relative_z='above',
+                                                   **kwargs)
+
+    def min_dz_biased_below(self, cube, sites, **kwargs):
+        '''
+        Return local grid neighbour with minimum vertical displacement,
+        biased to select grid points below the site altitude.
+
+        '''
+        return self.minimum_height_error_neighbour(cube, sites,
+                                                   relative_z='below',
+                                                   **kwargs)
+
+    def min_dz_land_no_bias(self, cube, sites, **kwargs):
+        '''
+        Return local grid neighbour with minimum vertical displacement.
+        Require land point neighbour if site is a land point.
+
+        '''
+        return self.minimum_height_error_neighbour(cube, sites,
+                                                   relative_z=None,
+                                                   land_constraint=True,
+                                                   **kwargs)
+
+    def min_dz_land_biased_above(self, cube, sites, **kwargs):
```

Comment:
I'm implementing this in the process function to allow for some of Gavin's requests as well, we'll see how well it works out.